### PR TITLE
kingpin docs, `noms blob [put | get]`

### LINKED ISSUE #2157: proposal: edit noms from cli: `noms struct` and friends
It would be really useful to make minor edits to data from the command line. Basically, to be able to do the Noms equivalent of `UPDATE stuffs SET foo=bar WHERE ...`.

Proposal:

```
noms struct [command] [args]

`noms struct` creates or modifies struct values.

Commands

    new <db> [--name=<name>] [key, value, key, value, ...]
    set <source> [--at=<rel-path>] <key> <val> [key, value, key, value, ...]
    del <source> [--at=<rel-path>] <key> [key, key, ...]
    name <source> [--at=<rel-path>] <name>

    <db>     : Database spec to create new struct in
    <name> : Name to assign to struct
    <source>  : Path to the root of the tree to modify. If [rel-path] is not specified, this will be the path to a struct to be modified. Otherwise, it will be one of that struct's parents.
    <rel-path> : An optional relative path inside <source> to the struct to be modified.
    <key>     : Key to add or update
    <val>     : Value to set key to

Example

    # Mark crime 14235 as resolved - this returns a new commit object with the specified
    change.

    > noms struct set http://localhost:8000::sf-crime.value [14235] STATUS "RESOLVED"
    Result: #kenik8esfarr4r0im46gl76o0edjkm4o

    > noms show http://localhost:8000::#kenik8esfarr4r0im46gl76o0edjkm4o
    struct Commit {

      ...

      14235: {
        TYPE: "stolen property",
        DESCRIPTION: "Missing bicycle",
        ...
        STATUS: "RESOLVED",
      }

     ...
    })

Formatting values

    Values are formatted as in the Human-Readable Serialization:
    - If a value is prefixed with @, it is a absolute path in this database to a value
    - If a value is wrapped in double-quotes, it is a string
    - If a value is prefixed with #, it's a hash
    - If the value is a bare `true` or `false`, it's a boolean
    - Otherwise the value is a number
```

```
noms map [command] [args]

`noms map` creates or modifies map values.

Commands

    new <kv...>
    set <source> [--at=<rel-path>] <kv...>
    del <source> [--at=<rel-path>] <k...>
```

```
noms list [command] [args]

`noms list` creates or modifies list values.

Commands

    new [val...]
    insert <source> [--at=<rel-path>] [pos] <val...>
    del <source> [--at=<rel-path>] pos [len]
```

```
noms set [command] [args]

`noms set` creates or modifies set values.

Commands

    new [val...]
    insert <source> [--at=<rel-path>] <val...>
    del <source> [--at=<rel-path>] [len]
```

```
noms blob [command] [args]

`noms blob` imports or outputs blobs from Noms

Commands

    import <url-or-file>
    export <noms-path>
   # Modification commands could come later if needed
```

### PR DESCRIPTION
**Pull request by jesseditson**
Fixes #2157:

Alternative/path to #3618

Summary
---

- removes `util/help.go` in favor of [kingpin](https://github.com/alecthomas/kingpin), which now handles printing help
- adds `noms blob import` and `noms blob export` commands, who's functionality is identical to the prior `blob-put` and `blob-get` samples, which they supercede
- adds [shell completion](https://github.com/alecthomas/kingpin#bashzsh-shell-completion) to noms (needs docs)

Unlike #3618, this only adds kingpin documentation and argument validation, rather than refactoring the commands.

That may still be too much of a change, as the validation will cause things like `--bool-flag=false` to fail. See the updated tests to see where this changed.

The reasoning behind this approach rather than trying to keep both around is that the `help` command would be missing the blob command without switching over - alternatively I could hardcode the blob docs in for now, but this seemed like a good path forward to me.

examples
---

output from `noms` and `noms help merge` are about the same as #3618

**output of `noms help blob`**
```
usage: noms blob <command> [<args> ...]

interact with blobs in a dataset

Flags:
  --help  Show context-sensitive help (also try --help-long and --help-man).

Subcommands:
  blob import [<flags>] <url-or-file> <dataset>
    imports a blob to a dataset

  blob export [<flags>] <dataset> [<file>]
    exports a blob from a dataset
```

**output of `noms help blob import`**
```
usage: noms blob import [<flags>] <url-or-file> <dataset>

imports a blob to a dataset

Flags:
      --help                   Show context-sensitive help (also try --help-long and --help-man).
  -v, --verbose                show more
  -q, --quiet                  show less
      --cpuprofile=CPUPROFILE  write cpu profile to file
      --memprofile=MEMPROFILE  write memory profile to file
      --blockprofile=BLOCKPROFILE
                               write block profile to file
      --concurrency=8          number of concurrent HTTP calls to retrieve remote resources

Args:
  <url-or-file>  a url or file to import
  <dataset>      the path to import to
```

**output of `noms help blob export`**
```
usage: noms blob export [<flags>] <dataset> [<file>]

exports a blob from a dataset

Flags:
      --help                   Show context-sensitive help (also try --help-long and --help-man).
  -v, --verbose                show more
  -q, --quiet                  show less
      --cpuprofile=CPUPROFILE  write cpu profile to file
      --memprofile=MEMPROFILE  write memory profile to file
      --blockprofile=BLOCKPROFILE
                               write block profile to file

Args:
  <dataset>  the dataset to export
  [<file>]   an optional file to save the blob to
```

### CODE REVIEW

**Comment by aboodman** [comment]
WDYT about `noms blob get|put` instead of `import|export`?

**Comment by jesseditson** [comment]
@aboodman piece of cake to change if you like, they both make sense to me!

File: cmd/noms/noms.go [hunk 1]
```diff
@@ -31,6 +33,13 @@ var commands = []*util.Command{
 	nomsVersion,
 }
 
+type kCommandHandler func(input string) (exitCode int)
```
Comment (aboodman) [inline cmd/noms/noms.go]: How about `command` and `commandHandler`. The `k` prefix is frequently used as a nemonic to mean 'constant', so it's confusing to see it here.

File: go/util/profile/profile.go [hunk 2]
```diff
@@ -30,6 +35,25 @@ func RegisterProfileFlags(flags *flag.FlagSet) {
 	}
 }
 
+func KAddProfileFlags(cmd *kingpin.CmdClause) {
```
Comment (aboodman) [inline go/util/profile/profile.go]: Same thing with the naming here, the `k` prefix is confusing. Can we just drop it?

File: cmd/noms/noms.go [hunk 3]
```diff
@@ -86,7 +105,128 @@ func main() {
 			return
 		}
 	}
+}
+
+var (
```
Comment (aboodman) [inline cmd/noms/noms.go]: The globals here skeeve me out. How about having caller just track these pointers itself?

File: cmd/noms/noms_blob.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	return blob, func(input string) int {
+		applyVerbosity()
+		profile.KApplyProfileFlags()
+		switch input {
+		case blobImport.FullCommand():
+			return nomsBlobImport(*importFile, *importDs, *concurrency)
+		case blobExport.FullCommand():
+			return nomsBlobExport(*exportDs, *exportPath)
+		}
+		// unreachable
```
Comment (aboodman) [inline cmd/noms/noms_blob.go]: We usually do `d.Panic("notreached")` in this case to make sure it doesn't happen.

**Comment by aboodman** [comment]
OK let's do get/put instead. Few characters shorter :).

**Comment by jesseditson** [comment]
done!

File: cmd/noms/noms.go (hunk 1 above)
Comment (jesseditson) [inline cmd/noms/noms.go]: yeah was wondering if this was gonna look too hungarian. Was planning on removing once the kingpin stuff no longer needed to be differentiated, but I think this is a simple enough file to not need that.

File: cmd/noms/noms.go (hunk 3 above)
Comment (jesseditson) [inline cmd/noms/noms.go]: Sounds good!

**Comment by jesseditson** [comment]
@aboodman feedback addressed, ptal!

**Comment by jesseditson** [comment]
The potentially controversial things I can think of are:

- can no longer do `--bool-flag=true`, must use `--bool-flag` instead (only an issue because it could break scripts)
- adds some dependencies (which is why this PR is 14k lines)
- kingpin uses a fluent API, which I haven't seen a ton of in go land (but I enjoy elsewhere)

but I think we've addressed those! I'll leave open for a bit and merge later or tomorrow

**Comment by aboodman** [comment]
No need to wait. It passes tests, land it!

**Comment by aboodman** [comment]
🚢 :shipit: 🚀 🎸 ▶️

**Pull request merged**